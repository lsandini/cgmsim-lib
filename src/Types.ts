import { TypeDateISO } from './TypeDateISO';
import { drugs } from './drug';

/**
 * Represents the direction of blood glucose change.
 */
export type Direction =
	| 'DoubleDown'
	| 'SingleDown'
	| 'FortyFiveDown'
	| 'Flat'
	| 'FortyFiveUp'
	| 'SingleUp'
	| 'DoubleUp';

/**
 * Represents drug information.
 */
export type Drug = {
	/** The timestamp of the drug entry. */
	time: number;
	/** The name of the drug. */
	drug: string;
	/** Additional notes for the drug. */
	notes: string;
	/** The amount of insulin associated with the drug. */
	insulin: number;
	/** Placeholder property for empty space. */
	empty_space: any;
};

/**
 * Represents a note with associated notes.
 */
export type Note = {
	/** The type of note. */
	type: 'Note';
	/** The content of the note. */
	notes: string;
};

/**
 * Represents data related to physical activity.
 */
export type Activity = {
	/** The number of steps taken (optional). */
	steps?: number;
	/** The heart rate during the activity (optional). */
	heartRate?: number;
	/** The timestamp indicating when the activity data was recorded in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents a blood glucose entry.
 */
export type Sgv = {
	/** The timestamp of the blood glucose entry in milliseconds. */
	mills: number;
	/** The blood glucose value. */
	sgv: number;
};

/**
 * Represents a blood glucose entry value type.
 */
export type EntryValueType = {
	/** The blood glucose value. */
	sgv: number;
	/** The direction of blood glucose change. */
	direction: string;
};

/**
 * Represents a blood glucose entry with additional date information.
 */
export type Entry = EntryValueType & {
	/** The timestamp of the entry. */
	date: number;
	/** The date in string format. */
	dateString: string;
	/** The type of entry. */
	type: 'sgv';
};

/**
 * Represents parameters for a profile.
 */
export type ProfileParams = {
	/** The basal insulin rate. */
	basal: number | { value: number; time: string; timeAsSecond?: number }[];
};

/**
 * Represents a profile with associated profile parameters.
 */
export type NSProfile = {
	/** The start date of the profile. */
	startDate: string;
	/** The default profile name. */
	defaultProfile: string;
	/** The store of profile parameters. */
	store: {
		[profileName: string]: ProfileParams;
	};
};

//Note', 'Profile Switch', 'Temp Basal', 'Meal Bolus', 'Bolus Wizard', 'Correction Bolus', 'Temporary Target', 'Insulin Change', 'Site Change', 'Sensor Start', 'Sensor Stop'

/**
 * Represents the treatment information for a Meal Bolus event.
 */
export type MealBolusTreatment = {
	/** The type of event, set to 'Meal Bolus'. */
	eventType:
		| 'Meal Bolus'
		| 'Bolus'
		| 'Correction Bolus'
		| 'Bolus Wizard'
		| 'Carb Correction';
	/** The amount of insulin administered for the meal bolus. */
	insulin?: number;
	/** The number of carbohydrates consumed. */
	carbs?: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents the treatment information for a Profile Switch event.
 */
export type ProfileSwitchTreatment = {
	/** The type of event, set to 'Profile Switch'. */
	eventType: 'Profile Switch';
	/** The duration of the profile switch in minutes. */
	duration: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
	/** The JSON representation of the new profile. */
	profileJson: string;
	/** The percentage change applied during the profile switch. */
	percentage: number;
};

/**
 * Represents the treatment information for a Temporary Basal event.
 */
export type TempBasalTreatment = {
	/** The type of event, set to 'Temp Basal'. */
	eventType: 'Temp Basal';
	/** The absolute insulin rate for the temporary basal. */
	absolute: number;
	/** The duration of the temporary basal in minutes. */
	duration: number;
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
};

/**
 * Represents the announcement as treatment.
 */
export type AnnouncementTreatment = {
	/** The date of the treatment creation in ISO format. */
	created_at: TypeDateISO;
	/** The type of event, set to 'Announcement'. */
	eventType: 'Announcement';
	/** Additional notes for the treatment. */
	notes: string;
};

/**
 * Represents treatment data.
 */
export type NSTreatment =
	| MealBolusTreatment
	| ProfileSwitchTreatment
	| TempBasalTreatment
	| AnnouncementTreatment;

export const isMealBolusTreatment = (
	treatment: NSTreatment,
): treatment is MealBolusTreatment =>
	treatment.eventType === 'Meal Bolus' ||
	treatment.eventType === 'Bolus' ||
	treatment.eventType === 'Bolus Wizard' ||
	treatment.eventType === 'Correction Bolus';

export const isAnnouncementTreatment = (
	treatment: NSTreatment,
): treatment is AnnouncementTreatment => treatment.eventType === 'Announcement';

export type NSTreatmentParsed = {
	drug: string;
	units: number;
	minutesAgo: number;
};

/**
 * Represents a treatment delta with additional minutes ago information.
 */
export type TreatmentBiexpParam = {
	units: number;
	minutesAgo: number;
	duration: number;
	peak: number;
};

/**
 * Represents a gender type.
 */
export type GenderType = 'Male' | 'Female';

/**
 * Represents parameters for a CGM simulation.
 */
export type CGMSimParams = {
	basalActivity: number;
	liverActivity: number;
	carbsActivity: number;
	bolusActivity: number;
	cortisoneActivity: number;
	alcoholActivity: number;
};

/**
 * Represents user parameters.
 */
export type UserParams = {
	nsUrl: string;
};

type PatientInfoBase = {
	/** Weight of the simulated user (in cm). */
	WEIGHT: number;
	/** Age of the simulated user. */
	AGE: number;
	/** Gender of the simulated user ('Male' or 'Female'). */
	GENDER: GenderType;
};

/**
 * Represents patient parameters for a cgmsim simulation.
 */
export type PatientInfoUva = PatientInfoBase;

/**
 * Represents patient parameters for a cgmsim simulation.
 */

export type PatientInfoCgmsim = PatientInfoBase & {
	/** Carbohydrate Ratio (CR) for insulin calculation. */
	CR: number;
	/** Insulin Sensitivity Factor (ISF) for insulin calculation. */
	ISF: number;
	/** Time taken for carbohydrates to be absorbed (in minutes default 360). */
	CARBS_ABS_TIME: number;
	/** Time period for insulin activity (Time Peak) (in minutes default 75). */
	TP: number;
	/** Duration of Insulin Action (DIA) (in hours default: 6). */
	DIA: number;
};

type MainParamsBase = {
	patient: PatientInfoBase;
	treatments: NSTreatment[];
	profiles: NSProfile[];
	pumpEnabled: boolean;
	activities?: Activity[];
	user: UserParams;
};
/**
 * Represents main parameters for a UVA simulation.
 */
export type MainParamsUVA = MainParamsBase & {
	lastState: UvaPatientState;
};

/**
 * Represents main parameters for a simulation.
 */
export type MainParams = MainParamsBase & {
	patient: PatientInfoCgmsim;
	entries: Sgv[];
};

/**
 * Represents the result of a simulation.
 */
export type SimulationResult = {
	sgv: number;
	deltaMinutes: number;
	carbsActivity: number;
	cortisoneActivity: number;
	basalActivity: number;
	bolusActivity: number;
	liverActivity: number;
	activityFactor: number;
	alcoholActivity: number;
	isf: { dynamic: number; constant: number };
};

/**
 * Represents the output of a UVA simulation.
 */
export type UvaOutput = {
	/** Glucose in plasma. */
	Gp: number;
	/** Glucose. */
	G?: number;
};

/**
 * Represents user parameters for a UVA simulation.
 */
export type UvaUserParams = {
	/** Insulin infusion rate. */
	iir: number;
	/** Insulin bolus amount. */
	ibolus: number;
	/** Carbohydrate intake. */
	carbs: number;
	/** Intensity of the simulation. */
	intensity: number;
};

/**
 * Represents a time delta for a UVA simulation.
 */
export type UvaDelta = 1;

/**
 * Represents an interval for a UVA simulation.
 */
export type UvaInterval = 5;

/**
 * Represents the state of a UVA patient.
 */
export type UvaPatientState = {
	/** Glucose in plasma. */
	Gp: number;
	/** Glucose in tissue. */
	Gt: number;
	/** Insulin delay compartment 1. */
	I_: number;
	/** Insulin in liver. */
	Il: number;
	/** Insulin in plasma. */
	Ip: number;
	/** Subcutaneous insulin in compartment 1. */
	Isc1: number;
	/** Subcutaneous insulin in compartment 2. */
	Isc2: number;
	/** Glucose mass in intestine. */
	Qgut: number;
	/** Carbs in stomach, solid phase. */
	Qsto1: number;
	/** Carbs in stomach, liquid phase. */
	Qsto2: number;
	/** Insulin in the interstitial fluid. */
	X: number;
	/** Insulin delay compartment 2. */
	XL: number;
	/** Increased energy consumption estimated through heart rate. */
	Y: number;
	/** Exercise-induced changes in insulin sensitivity. */
	Z: number;
	/** W. */
	W: number;
};

/**
 * Represents parameters for a UVA simulation.
 */
export type UvaPatientType = {
	/** Body weight. */
	BW: number;
	/** Steady-state of glucose in plasma. */
	Gpeq: number;
	/** Heart rate at rest. */
	HRb: number;
	/** Maximal heart rate (220-age). */
	HRmax: number; //220-age
	/** Distribution volume of glucose. */
	VG: number;
	/** Rate parameter from Gp to Gt. */
	k1: number;
	/** Rate parameter from Gt to Gp. */
	k2: number;
	/** Distribution volume of insulin. */
	VI: number;
	/** Rate parameter from Il to Ip. */
	m1: number;
	/** Rate parameter from Ip to Il. */
	m2: number;
	/** Rate parameter from Ip to periphery. */
	m4: number;
	/** Rate parameter of hepatic extraction (slope). */
	m5: number;
	/** Rate parameter of hepatic extraction (offset). */
	m6: number;
	/** Steady-state hepatic extraction of insulin. */
	HEeq: number;
	/** Maximal emptying rate of stomach. */
	kmax: number;
	/** Minimal emptying rate of stomach. */
	kmin: number;
	/** Rate constant of intestinal absorption. */
	kabs: number;
	/** Rate of grinding. */
	kgri: number;
	/** Fraction of intestinal absorption. */
	f: number;
	/** Extrapolated at zero glucose and insulin. */
	kp1: number;
	/** Liver glucose effectiveness. */
	kp2: number;
	/** Amplitude of insulin action on the liver. */
	kp3: number;
	/** Amplitude of portal insulin action on the liver. */
	kp4: number;
	/** Delay between insulin signal and insulin action. */
	ki: number;
	/** Glucose uptake by the brain and erythrocytes. */
	Fcns: number;
	/** Michaelis-Menten constant (offset). */
	Vm0: number;
	/** Michaelis-Menten constant (slope). */
	Vmx: number;
	/** Michaelis-Menten constant (offset). */
	Km0: number;
	/** Insulin action on the peripheral glucose utilization. */
	p2u: number;
	/** Glomerular filtration rate. */
	ke1: number;
	/** Renal threshold of glucose. */
	ke2: number;
	/** Rate constant of nonmonomeric insulin absorption. */
	ka1: number; // [Dalla Man, JDST, 2007]
	/** Rate constant of monomeric insulin absorption. */
	ka2: number; // [Dalla Man, JDST, 2007]
	/** Rate constant of insulin dissociation. */
	kd: number;
	/** Factor for exercise-induced increase in insulin sensitivity. */
	A: number; // [Dalla Man, JDST, 2009]
	/** Factor for exercise-induced increase in insulin-independent glucose clearance. */
	beta: number; // [Dalla Man, JDST, 2009]
	/** Factor for exercise-induced increase in glucose uptake. */
	gamma: number; // [Dalla Man, JDST, 2009]
	/** Parameter for calculating Z. */
	a: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Y. */
	Thr: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Z. */
	Tin: number; // [Dalla Man, JDST, 2009]
	/** Time constant for Z. */
	Tex: number; // [Dalla Man, JDST, 2009]
	/** Parameter for calculating Z. */
	n: number; // [Dalla Man, JDST, 2009]
};
