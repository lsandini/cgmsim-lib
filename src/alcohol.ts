import { TreatmentDelta, Treatment, TreatmentDrug, GenderType } from './Types';
import { getDrugActivity } from './drug';
import logger, { getDeltaMinutes, getTreatmentActivity } from './utils';

function getAlcoholActivity(
	gender: 'Male' | 'Female',
	weightKg: number,
	peakMin: number,
	durationMin: number,
	timeMin: number, //minutesAgo
	units: number,
): number {
	const peakAlcoholAmountMin = 60; //min
	const r = gender === 'Male' ? 0.68 : 0.55;
	const eliminationRate = gender === 'Male' ? 0.016 / 60 : 0.018 / 60; //g/100ml/min
	const peakAmount = ((units * 12) / (weightKg * 1000 * r)) * 100; //g/100ml
	const washoutDuration = peakAlcoholAmountMin + peakAmount / eliminationRate; //min
	const unitsWeighted = units * 12 * (80 / weightKg);
	if (washoutDuration < timeMin) {
		return (
			getTreatmentActivity(
				peakMin,
				durationMin,
				timeMin - washoutDuration,
				unitsWeighted,
			) / 35
		);
	}
	return 0;
}

export const computeAlcoholActivity = (
	treatments: TreatmentDelta[],
	weight: number,
	gender: GenderType,
) => {
	const treatmentsActivity = treatments.map((e) => {
		const minutesAgo = e.minutesAgo;
		const units = e.units;
		const activity = getAlcoholActivity(
			gender,
			weight,
			e.peak,
			e.duration,
			minutesAgo,
			units,
		);
		return activity;
	});
	logger.debug('these are the last Alcohol: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct;
};

export default function (
	treatments: TreatmentDrug[],
	weight: number,
	gender: GenderType,
): number {
	//Find Alcohol boluses

	const lastALC = getDrugActivity(treatments, weight, 'ALC');
	const activityALC =
		lastALC.length > 0 ? computeAlcoholActivity(lastALC, weight, gender) : 0;
	logger.debug('these are the last ALC: %o', { lastALC, activityALC });

	return activityALC / 2;
}
