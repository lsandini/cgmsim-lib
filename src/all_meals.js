const moment = require('moment');
module.exports = function(treatments = []) {
  const now = moment();
  const meals = treatments
    .filter(e => e.carbs && moment(e.mills).diff(now, 'minutes') > -360)
    .map(e => ({
      ...e,
      time:  (Date.now() - moment(e.mills).valueOf()) / (1000 * 60)
    }));

  console.log('Last 6 hours meal:', meals);
  return meals;
};
