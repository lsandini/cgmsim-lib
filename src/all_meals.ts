import logger from './utils';
import { Treatment, TreatmentDelta } from './Types';
import { getDeltaMinutes } from './utils';

//const logger = pino();

export default function (treatments: Treatment[]): TreatmentDelta[] {

	const meals = treatments
		.filter(e => e.carbs && getDeltaMinutes(e.mills) > 360)
		.map(e => ({
			...e,
			minutesAgo: getDeltaMinutes(e.mills),
		}));

	logger.info('Last 6 hours meal:', meals);
	return meals;
}

