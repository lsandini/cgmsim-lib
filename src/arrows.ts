import logger from './utils';
import { Sgv, Direction } from './Types';

//const logger = pino();


export default function (sgvs: Pick<Sgv, 'sgv'>[]): { sgvdir: number; direction: Direction }[] {
	const arrows = [];
	if (sgvs && sgvs.length >= 4) {
		const sgvdir1 = sgvs[0].sgv - sgvs[1].sgv;
		const sgvdir2 = sgvs[1].sgv - sgvs[2].sgv;
		const sgvdir3 = sgvs[2].sgv - sgvs[3].sgv;
		const sgvdir15min = (sgvdir1 + sgvdir2 + sgvdir3) / 3;
		logger.info('this is the mean SGV 5 min variation in the last 15 minutes: %o', sgvdir15min, 'mg/dl');

		if (sgvdir15min < -10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleDown'
			});

		} else if (sgvdir15min < -6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleDown'
			});

		} else if (sgvdir15min < -2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveDown'
			});

		} else if (sgvdir15min < 2) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'Flat'
			});

		} else if (sgvdir15min < 6) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'FortyFiveUp'
			});

		} else if (sgvdir15min < 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'SingleUp'
			});

		} else if (sgvdir15min >= 10) {
			arrows.push({
				sgvdir: sgvdir15min,
				direction: 'DoubleUp'
			});
		}

	} else {
		arrows.push({
			sgvdir: 0,
			direction: 'Flat'
		});
	}
	logger.info(arrows);

	return arrows;
}