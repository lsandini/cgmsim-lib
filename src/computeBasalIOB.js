const moment = require('moment');
module.exports = function(treatments) {

	//Find basal boluses
    const basal = treatments && treatments.length ? treatments.filter(e => e.notes && e.insulin && e.drug !== 'hum')
        .map(e => ({
            time: e.created_at,
            notes: e.notes,
            insulin: e.insulin,
            empty_space: e.notes.indexOf(' ')
        })) : [];

    const basals = basal.map(entry => ({
        ...entry,
        time: moment(entry.time).valueOf()
    }));
    const timeSinceBasalMin = basals.map(entry => ({
        ...entry,
        time: (Date.now() - moment(entry.time).valueOf()) / (1000 * 60 * 60),
        drug: entry.notes.slice(0, 3),
        //		insulin: parseInt(entry.notes.slice(entry.empty_space), 10)
    }));

    const lastBasals = timeSinceBasalMin.filter(function(e) {
        return e.time <= 45; // keep only the basals from the last 45 hours
    });

    const lastGLA = lastBasals.filter(function(e) {
        return e.drug === 'gla' || e.drug === 'Gla' || e.drug === 'lan' || e.drug === 'Lan'; // keep only the glas from the last 45 hours
    });

    const lastDET = lastBasals.filter(function(e) {
        return e.drug === 'det' || e.drug === 'Det' || e.drug === 'lev' || e.drug === 'Lev'; // keep only the dets from the last 45 hours
    });

    const lastTOU = lastBasals.filter(function(e) {
        return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
    });

    const lastDEG = lastBasals.filter(function(e) {
        return e.drug === 'deg' || e.drug === 'Deg' || e.drug === 'tre' || e.drug === 'Tre'; // keep only the degs from the last 45 hours
    });

    return {
        lastDET,
        lastGLA,
        lastTOU,
        lastDEG,
    };
};