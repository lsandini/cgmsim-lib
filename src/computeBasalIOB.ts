import { TreatmentDelta, Treatment } from './Types'
import logger, { getDeltaMinutes, getInsulinActivity } from './utils';


export const computeBasalActivity = (treatments: (TreatmentDelta & { duration: number, peak: number })[]) => {
	// activities be expressed as U/min !!!
	const treatmentsActivity = treatments.map(e => {
		const minutesAgo = e.minutesAgo;
		const insulin = e.insulin;
		return getInsulinActivity(e.peak, e.duration, minutesAgo, insulin);		
	});
	logger.info('these are the last Slow INSULINS: %o', treatmentsActivity);
	const resultAct = treatmentsActivity.reduce((tot, activity) => {
		return tot + activity;
	}, 0);
	return resultAct
}

export default function (treatments: Treatment[], weight: number): number {
	//Find basal boluses
	const basals = treatments && treatments.length ?
		treatments.filter(e => e.notes)
			.map(e => {
				const lastIndexEmptySpace = e.notes.lastIndexOf(' ');
				return {
					...e,
					minutesAgo: getDeltaMinutes(e.created_at),
					drug: e.notes.slice(0, 3),
					// insulin: parseInt(e.notes.slice(-2)) 
					insulin: parseInt(e.notes.slice(lastIndexEmptySpace), 10)
				}
			}) : [];

	const lastBasals = basals.filter(function (e) {
		return e.minutesAgo <= 45 * 60; // keep only the basals from the last 45 hours
	});

	const lastGLA = lastBasals
		.filter((e) => {
			return e.drug === 'gla' || e.drug === 'Gla' || e.drug === 'lan' || e.drug === 'Lan'; // keep only the glas from the last 45 hours
		}).map(e => {
			const duration = (22 + (12 * e.insulin / weight)) * 60;
			const peak = (duration / 2.5);
			return {
				...e,
				duration,
				peak,
			}
		});
	logger.debug('these are the last GLA: %o', lastGLA);
	const activityGLA = computeBasalActivity(lastGLA)
	logger.info('these is the total GLA activity: %o', activityGLA);

	const lastDET = lastBasals.filter(function (e) {
		return e.drug === 'det' || e.drug === 'Det' || e.drug === 'lev' || e.drug === 'Lev'; // keep only the dets from the last 45 hours
	}).map(e => {
		const duration = (14 + (24 * e.insulin / weight)) * 60;
		const peak = (duration / 3);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last DET: %o', lastDET);
	const activityDET = computeBasalActivity(lastDET)
	logger.info('these is the total DET activity: %o', activityDET);

	const lastTOU = lastBasals.filter(function (e) {
		return e.drug === 'tou' || e.drug === 'Tou'; // keep only the tous from the last 45 hours
	})
	.map(e => {
		const duration = (24 + (14 * e.insulin / weight)) * 60;
		const peak = (duration / 2.5);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last TOU: %o', lastTOU);
	const activityTOU = computeBasalActivity(lastTOU)
	logger.info('these is the total TOU activity: %o', activityTOU);

	const lastDEG = lastBasals.filter(function (e) {
		return e.drug === 'deg' || e.drug === 'Deg' || e.drug === 'tre' || e.drug === 'Tre'; // keep only the degs from the last 45 hours
	}).map(e => {
		const duration = 42 * 60;
		const peak = (duration / 3);
		return {
			...e,
			duration,
			peak,
		}
	});
	logger.debug('these are the last deg: %o', lastDEG);
	const activityDEG = computeBasalActivity(lastDEG);
	logger.info('these is the total deg activity: %o', activityDEG);

	const result = {
		lastDET,
		lastGLA,
		lastTOU,
		lastDEG,		
	};
	return activityDEG + activityDET + activityGLA + activityTOU;
}