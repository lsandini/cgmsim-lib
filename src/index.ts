import simulator from './CGMSIMsimulator';
import simulatorUVA from './UVAsimulator';
import downloads from './downloads';
import uploadEntries from './upload-entries';
import uploadActivity from './upload-activity';
import loadActivity from './load-activity';
import uploadNotes from './upload-notes';
import arrows from './arrows';


export { arrows, simulator, downloads, uploadEntries, uploadActivity, loadActivity, simulatorUVA, uploadNotes };
