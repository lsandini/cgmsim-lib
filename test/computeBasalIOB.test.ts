import { treatments } from "./inputTest";
import computeBasalIOB from "../src/computeBasalIOB";
import { Treatment } from "../src/Types";

describe('test computeBasalIOB', () => {

	beforeAll(() => {
		jest.useFakeTimers('modern');
		jest.setSystemTime(new Date('2022-05-07T11:20:00'));
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	test('detection drug', () => {
		const result = computeBasalIOB(treatments as unknown as Treatment[],80);

		expect(result).toMatchSnapshot()
	})
})