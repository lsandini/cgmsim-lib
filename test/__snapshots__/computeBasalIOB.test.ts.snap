// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test computeBasalIOB detection drug 1`] = `0.005679895517704191`;
